src/repr/src/row/collection.rs

```diff
@@ -225,32 +332,163 @@ impl RustType<ProtoEncodedRowMetadata> for EncodedRowMetadata {
     }
 }
 
-/// Provides a sorted view of a [`RowCollection`].
+/// Used to store a [`RowCollection`] in a [`BinaryHeap`] without creating any references.
+/// Each row in the row collection is successively "inserted" into the heap by simply updating
+/// `row_index`. Once all entries are exhausted, the `HeapRowCollection` is removed from the heap.
+///
+/// Implements a custom [`Ord`] to make the heap a _min-heap_.
+#[derive(Clone)]
+pub struct HeapRowCollection {
+    /// A sorted [`RowCollection`].
+    row_collection: SortedRowCollection,
+    /// The `row_collection` index that is considered active in the heap.
+    row_index: usize,
+    /// The number of diffs left to process of the current row.
+    diffs_left: usize,
+    /// Sort order
+    sort_by: Arc<Mutex<dyn FnMut(&RowRef, &RowRef) -> Ordering + Sync + Send>>,
```

Comment:
The `Mutex` here allows interior mutability but doesn't really feel right. The tradeoffs are:
* We need a `FnMut` here to reuse `Vec` allocations across all comparisons.
* std::BinaryHeap (needed for a `N*log(k)` merge) needs an `Ord` comparison which does not have mutable access.
* Any type here needs to be `Send + Sync`, hence no `RefCell` etc.

Perhaps I'm missing a way to simplify this logic.